atomicops.h

```diff
@@ -581,6 +583,64 @@ namespace moodycamel
 		        }
 		    }
 		};
+#elif defined(__HAIKU__)
+		//---------------------------------------------------------
+		// Semaphore (Haiku)
+		//---------------------------------------------------------
+		class Semaphore
+		{
+		private:
+		    sem_id m_sema;
+
+		    Semaphore(const Semaphore& other);
+		    Semaphore& operator=(const Semaphore& other);
+
+		public:
+		    AE_NO_TSAN Semaphore(int initialCount = 0) : m_sema()
+		    {
+		        assert(initialCount >= 0);
+		        m_sema = create_sem(initialCount, "");
+		        assert(m_sema >= B_OK);
+		    }
+
+		    AE_NO_TSAN ~Semaphore()
+		    {
+		        delete_sem(m_sema);
+		    }
+
+		    bool wait() AE_NO_TSAN
+		    {
+		        // accquire the semaphore
+		        status_t rc;
+		        rc = acquire_sem(m_sema);
+		        return rc == B_NO_ERROR;
+		    }
+
+			bool try_wait() AE_NO_TSAN
+			{
+				// accquire the semaphore
+		        status_t rc;
+		        rc = acquire_sem_etc(m_sema, 1, B_RELATIVE_TIMEOUT, 0);
+		        return rc == B_NO_ERROR;
+			}
+
+			bool timed_wait(std::uint64_t usecs) AE_NO_TSAN
+			{
+		        status_t rc;
+		        rc = acquire_sem_etc(m_sema, 1, B_RELATIVE_TIMEOUT, (bigtime_t) usecs);
+		        return rc == B_NO_ERROR;
+			}
+
+		    void signal() AE_NO_TSAN
+		    {
+		        while (release_sem(m_sema) != B_NO_ERROR);
```

Comment:
Looks like an infinite loop on error? As far as I can tell from the docs, `release_sem` does not return spuriously.

Reply:
I removed the while loop as it didn't make sense. The reason I had it in there was because I copied the unix version of it and replaced the methods at first, which was kind of naive from me. If it is ok I'd create a recipe for HaikuPorts once this is merged so it can be installed as an develpment header from a package